src/psyclone/tests/psyir/symbols/datatype_test.py

```diff
@@ -890,44 +890,77 @@ def test_unsupported_fortran_type_replace_symbols():
+    flag2 = stype.lookup_component("flag2")
     assert isinstance(flag2, StructureType.ComponentType)
     assert flag2.initial_value.value == "1"
     with pytest.raises(TypeError) as err:
-        stype.add(1, "hello", "hello", None)
+        stype.add_component(1, "hello", "hello", None)
     assert ("name of a component of a StructureType must be a 'str' but got "
             "'int'" in str(err.value))
     with pytest.raises(TypeError) as err:
-        stype.add("hello", "hello", "hello", None)
+        stype.add_component("hello", "hello", "hello", None)
     assert ("type of a component of a StructureType must be a 'DataType' "
             "or 'DataTypeSymbol' but got 'str'" in str(err.value))
     with pytest.raises(TypeError) as err:
-        stype.add("hello", INTEGER_TYPE, "hello", None)
+        stype.add_component("hello", INTEGER_TYPE, "hello", None)
     assert ("visibility of a component of a StructureType must be an instance "
             "of 'Symbol.Visibility' but got 'str'" in str(err.value))
     with pytest.raises(TypeError) as err:
-        stype.add("hello", INTEGER_TYPE, Symbol.Visibility.PUBLIC, "Hello")
+        stype.add_component("hello", INTEGER_TYPE, Symbol.Visibility.PUBLIC,
+                            "Hello")
     assert ("The initial value of a component of a StructureType must be "
             "None or an instance of 'DataNode', but got 'str'."
             in str(err.value))
     with pytest.raises(KeyError):
-        stype.lookup("missing")
+        stype.lookup_component("missing")
     # Cannot have a recursive type definition
     with pytest.raises(TypeError) as err:
-        stype.add("hello", stype, Symbol.Visibility.PUBLIC, None)
+        stype.add_component("hello", stype, Symbol.Visibility.PUBLIC, None)
     assert ("attempting to add component 'hello' - a StructureType definition "
             "cannot be recursive" in str(err.value))
 
+    # Add a procedure component
+    stype.add_procedure_component("proc1", INTEGER_TYPE,
+                                  Symbol.Visibility.PUBLIC,
+                                  None)
+    proc1 = stype.lookup_procedure_component("proc1")
+    assert not proc1.initial_value
+    assert isinstance(proc1, StructureType.ComponentType)
+    with pytest.raises(TypeError) as err:
+        stype.add_procedure_component(None, None, None, None)
+    assert ("The name of a procedure component of a StructureType must "
+            "be a 'str' but got 'NoneType'" in str(err.value))
+    with pytest.raises(TypeError) as err:
+        stype.add_procedure_component("proc2", None, None, None)
+    assert ("The type of a procedure component of a StructureType must "
+            "be a 'DataType' but got 'NoneType'" in str(err.value))
+    with pytest.raises(TypeError) as err:
+        stype.add_procedure_component("proc3", StructureType(), None, None)
+    assert ("The type of a procedure component of a StructureType cannot "
```

Comment:
Does this mean the return type of the procedure component? If so, why isn't it permitted to be a structure?

Reply:
You're right, that's probably too restrictive. My issue what that StructureType <: DataType but in that case the DataTypeSymbol should be used instead of the StructureType

Reply:
In any case I'm really not sure if there's even a point in having the `datatype` info for a procedure...

Reply:
We do have types for Procedures (so a RoutineSymbol does have a datatype) so I think it makes sense.